test/navigation/imu_preprocessing.cpp

```diff
@@ -0,0 +1,68 @@
+#include <iostream>
+
+#include <gtest/gtest.h>
+
+#include <core/logger.hpp>
+#include <core/types.hpp>
+#include <navigation/preprocess_imu.hpp>
+#include <utils/manual_time.hpp>
+
+namespace hyped::test {
+
+core::Float epsilon = 0.001;
```

Comment:
epsilon is too large here. This would allow a 1-step rounding error of that size which would mess up our estimates quite quickly given the frequency of iterations when the pod is running.

Reply:
Ive now used core::kEpsilon, is that a better order of magnitude?

Reply:
Possibly not. Leave it for now and we can talk about it a bit this evening (also weirdly enough the ideas in the error analysis section of your numerical ode course might be of interest but not directly applicable...). A build up of error is one of the things we have to be really careful of in the navigation team. We should be fine on a 100m track but we should be aiming to reduce all errors to allow the potential scaling of our system at a later date.

Reply:
i commented in another place about how to do this from a code pov

as for error analysis: i think this would warrant something really cool. the HYPED could propose a stats or computing project doing this. i think the error problems combined with the computational limitations would make this really interesting. i think the project wouldn't require any actual C++ code as HYPED is probably better at doing that themselves. @kshxtij talk to the TDs? I'm happy to help with the proposal. Needs to be soon-ish, though.

Reply:
@mifrandir this is a complete problem across navigation. I don't think we currently have an explicit method for reducing the error since the closest thing we have to a direct way of viewing the state of the dynamic system is the wheel encoder displacement (although the kalman filter does it implicitly from what I remember). Currently the approach is more "check we haven't gone too far wrong" than considering the bounds on the errors or implementing any sort of correction/rigorous error reduction methods. This would probably fit quite well as a computational statistics problem (investigating Bayesian filters) and would probably work best if it was implemented in some sort of code (e.g. python, R, etc.) so that implementation is at least considered. Once we get the car up and running, we should be able to provide plenty of real world data for working with.

Reply:
i agree fully; let's propose something

i assume you know some people in the school that do this sort of stuff?

Reply:
I think this is a great idea, do we know what the timelines in SoM/SoI look like for self proposed projects?

I've heard SoI is March but I can't find the dates on a website anywhere (or atleast a website i can access as a 2nd year)

Reply:
SoI is 3 March, but we need a supervisor.